pkg/controllers/node/termination/machine_test.go

```diff
@@ -0,0 +1,512 @@
+		cloudProvider.CreatedNodeClaims[node.Spec.ProviderID] = nodeclaimutil.New(machine)
+		queue.Reset()
+	})
+
+	AfterEach(func() {
+		ExpectCleanedUp(ctx, env.Client)
+		fakeClock.SetTime(time.Now())
+		cloudProvider.Reset()
+
+		// Reset the metrics collectors
+		metrics.NodesTerminatedCounter.Reset()
+		termination.TerminationSummary.Reset()
+	})
+
+	Context("Reconciliation", func() {
+		It("should delete nodes", func() {
+			ExpectApplied(ctx, env.Client, node)
+			Expect(env.Client.Delete(ctx, node)).To(Succeed())
+			node = ExpectNodeExists(ctx, env.Client, node.Name)
+			ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(node))
+			ExpectNotFound(ctx, env.Client, node)
+		})
+		It("should delete machines associated with nodes", func() {
+			ExpectApplied(ctx, env.Client, node, machine)
+			Expect(env.Client.Delete(ctx, node)).To(Succeed())
+			node = ExpectNodeExists(ctx, env.Client, node.Name)
+			ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(node))
+			ExpectExists(ctx, env.Client, machine)
+			ExpectFinalizersRemoved(ctx, env.Client, machine)
+			ExpectNotFound(ctx, env.Client, node, machine)
+		})
+		It("should not race if deleting nodes in parallel", func() {
+			var nodes []*v1.Node
+			for i := 0; i < 10; i++ {
+				node = test.Node(test.NodeOptions{
+					ObjectMeta: metav1.ObjectMeta{
+						Finalizers: []string{v1alpha5.TerminationFinalizer},
+					},
+				})
+				ExpectApplied(ctx, env.Client, node)
+				Expect(env.Client.Delete(ctx, node)).To(Succeed())
+				node = ExpectNodeExists(ctx, env.Client, node.Name)
+				nodes = append(nodes, node)
+			}
+
+			var wg sync.WaitGroup
+			// this is enough to trip the race detector
+			for i := 0; i < 10; i++ {
+				wg.Add(1)
+				go func(node *v1.Node) {
+					defer GinkgoRecover()
+					defer wg.Done()
+					ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(node))
+				}(nodes[i])
+			}
+			wg.Wait()
+			ExpectNotFound(ctx, env.Client, lo.Map(nodes, func(n *v1.Node, _ int) client.Object { return n })...)
+		})
+		It("should exclude nodes from load balancers when terminating", func() {
+			// This is a kludge to prevent the node from being deleted before we can
+			// inspect its labels
+			podNoEvict := test.Pod(test.PodOptions{
+				NodeName: node.Name,
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations:     map[string]string{v1alpha5.DoNotEvictPodAnnotationKey: "true"},
+					OwnerReferences: defaultOwnerRefs,
+				},
+			})
+
+			ExpectApplied(ctx, env.Client, node, podNoEvict)
+
+			Expect(env.Client.Delete(ctx, node)).To(Succeed())
+			node = ExpectNodeExists(ctx, env.Client, node.Name)
+			ExpectReconcileSucceeded(ctx, terminationController, client.ObjectKeyFromObject(node))
+			node = ExpectNodeExists(ctx, env.Client, node.Name)
+			Expect(node.Labels[v1.LabelNodeExcludeBalancers]).Should(Equal("karpenter"))
+		})
```

Comment:
"Should evict pods that tolerate the karpenter disruption taint" to make sure that this edge case is covered for Machines